iio/iio_trigger.h

```diff
@@ -0,0 +1,153 @@
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef IIO_TRIGGER_H_
+#define IIO_TRIGGER_H_
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "iio.h"
+#include "iio_types.h"
+#include "no_os_irq.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define TRIG_MAX_NAME_SIZE 20
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+/**
+ * @struct iio_hw_trig
+ * @brief IIO hardware trigger structure
+ */
+struct iio_hw_trig {
+	/** IIO descriptor */
+	struct iio_desc	**iio_desc;
+	/** Interrupt descriptor to be linked with the trigger */
+	struct no_os_irq_ctrl_desc *irq_ctrl;
+	/** Interrupt id to be linked with the trigger */
+	uint32_t irq_id;
+	/** Interrupt trigger level */
+	enum no_os_irq_trig_level irq_trig_lvl;
+	/** Device trigger name */
+	char name[TRIG_MAX_NAME_SIZE + 1];
+};
+
+/**
+ * @struct iio_hw_trig_cb_info
+ * @brief Hardware trigger callback extra information structure
+ */
+struct iio_hw_trig_cb_info {
+	/** Platform specific event that triggers the calling of the callback. */
+	enum no_os_irq_event event;
+	/** Interrupt source peripheral specifier. */
+	enum no_os_irq_peripheral peripheral;
+	/** This will be used to store HAL specific descriptors */
+	void *handle;
+};
+
+/**
+ * @struct iio_hw_trig_init_param
+ * @brief IIO hardware trigger initialization structure
+ */
+struct iio_hw_trig_init_param {
+	/** IIO descriptor */
+	struct iio_desc **iio_desc;
+	/** Interrupt descriptor to be linked with the trigger */
+	struct no_os_irq_ctrl_desc *irq_ctrl;
+	/** Interrupt id to be linked with the trigger */
+	uint32_t irq_id;
+	/** Interrupt trigger level */
+	enum no_os_irq_trig_level irq_trig_lvl;
+	/** Additional interrupt callback information */
+	struct iio_hw_trig_cb_info cb_info;
+	/** Device trigger name */
+	const char *name;
+};
+
+/**
+ * @struct iio_sw_trig
+ * @brief IIO software trigger structure
+ */
+struct iio_sw_trig {
+	/** IIO descriptor */
+	struct iio_desc **iio_desc;
```

Comment:
Question: do we need a double pointer for the `iio_desc`? Could it work with only a simple pointer?

Reply:
a simple pointer will do the job, thanks! I made the updates.

Reply:
Actually, after executing some tests, a double pointer is needed for iio_desc.

Reason: iio_desc is used as a double pointer in iio_app_run_with_trigs, because the memory allocation is taking place in iio_init function from iio_app.c. Here, a new value is given to *iio_desc, see line 1821 from iio_app `*desc = ldesc`.
The new value, is no longer equal to the value we passed to trigger init functions leading to erroneous behavior.

However, the value of **iio_desc remains constant and can be used safely.